adam_core/utils/indexable.py

```diff
@@ -1,174 +1,451 @@
+        not all of the same length.
+
+        Sets
+        ----
+        self._member_length : int
+            The length of the members of the class.
+        self._member_index : `~numpy.ndarray`
+            The index of the members of the class.
+
+        Raises
+        ------
+        ValueError: If the members are not all of the same length.
+        """
+        # Scan members and if they are of a type that is sliceable
+        # then store the length. All sliceable members should have the same length
+        # along their first axis.
+        member_lengths = []
+        for k, v in self.__dict__.items():
+            if isinstance(v, SLICEABLE_DATA_STRUCTURES):
+                member_lengths.append(len(v))
 
+        if len(np.unique(member_lengths)) != 1:
+            raise ValueError("All sliceable members must have the same length.")
         else:
-            err = "index must be a str, numpy.ndarray"
-            raise ValueError(err)
+            member_length = member_lengths[0]
 
-        self._index = pd.MultiIndex.from_arrays(
-            [class_index, array_index], names=["class_index", "array_index"]
-        )
+        self._member_length = member_length
+        self._member_index = np.arange(0, member_length, dtype=int)
 
         return
 
-    def __getitem__(self, i: Union[int, slice, tuple, list, np.ndarray, pd.MultiIndex]):
-        ind = self._handle_index(i)
-        copy = deepcopy(self)
+    def set_index(self, index_values: Optional[Union[str, npt.ArrayLike]] = None):
+        """
+        Set an index on the class for the given values or attribute name.
 
-        for k, v in self.__dict__.items():
-            if isinstance(
-                v,
-                (np.ndarray, np.ma.masked_array, list, Time, Indexable, pd.MultiIndex),
-            ):
-                copy.__dict__[k] = v[ind]
-            elif isinstance(v, UNSLICEABLE_DATA_STRUCTURES):
-                copy.__dict__[k] = v
-            elif v is None:
-                pass
-            else:
-                err = f"{type(v)} are not supported."
-                raise NotImplementedError(err)
+        Parameters
+        ----------
+        values : str, np.ArrayLike
+            The values to be indexed. If a string is given then the values are taken from the
+            attribute of the same name. If an array is given then the values are taken from the
+            array itself. If None is given then the index is set to the range of length
+            of the class's sliceable members.
+
+        Sets
+        ----
+        self._class_index : `~numpy.ndarray`
+            The externally facing index of the class.
+        self._class_index_to_members : `~numpy.ndarray`
+            The mapping from the externally facing index to the class members.
+        self._class_index_to_members_is_slice : bool
+            Whether the mapping from the externally facing index to the class members are slices.
```

Comment:
This `..._is_slice` attribute sticks out to me. It shows that `self._class_index_to_members` is one variable being reused for two totally different purposes. It can be either used to build a slice-based index to unique values, or an integer mapping to non-integer values for attribute-based lookups. 

Maybe those should just be two different attributes: `self._class_index_unique_slices`, and `self._class_index_int_mapping`.

Even if you don't split it into two variables, I think `_class_index_to_members` is a very confusing name, since the values it holds are _not_ used as any sort of mapping to members. They are used as indices when looking inside `self._class_index`.

Reply:
Done, I've split it into two attributes. We can change the names in a future PR, but I have:
`class_index_to_slices` and `class_index_to_integers`. 

I'm actually thinking we can drop `member_index` at some point since that is always just an array of integers ranging from 0 to the length of the underlying members.

Reply:
I've removed the member_index altogether. Instead, Indexable._query_index will compute the full indices into the member arrays and then slice them according to the indices passed to the function.

Reply:
Oh, that's fantastic! Getting rid of `member_index` clarifies the code a lot. It seems that `_class_index` could just be renamed `_index` at this point, right?

Reply:
Yeah actually, let me go ahead a submit a follow-up PR with some additional clean up.